src/main/java/org/rumbledb/compiler/StaticContextVisitor.java

```diff
@@ -127,39 +127,58 @@ public StaticContext visitFunctionCall(FunctionCallExpression expression, Static
     // region FLWOR
     @Override
     public StaticContext visitFlowrExpression(FlworExpression expression, StaticContext argument) {
-        StaticContext result = this.visit(expression.getStartClause(), argument);
-        for (Clause clause : expression.getContentClauses()) {
+        Clause clause = expression.getReturnClause().getFirstClause();
+        StaticContext result = this.visit(clause, argument);
+        while (clause != null) {
             result = this.visit(clause, result);
+            clause = clause.getNextClause();
         }
-
-        result = this.visit(expression.getReturnClause(), result);
         return result;
     }
 
     // region FLWOR vars
     @Override
-    public StaticContext visitForClauseVar(ForClauseVar expression, StaticContext argument) {
+    public StaticContext visitForClause(ForClause clause, StaticContext argument) {
         // TODO visit at...
-        this.visit(expression.getExpression(), argument);
-        expression.initHighestExecutionAndVariableHighestStorageModes(this.visitorConfig);
+        this.visit(clause.getExpression(), argument);
+        clause.initHighestExecutionMode(this.visitorConfig);
+
+        StaticContext result = new StaticContext(argument);
+        // TODO for now we only support as/default, no inference, flags
+        result.addVariable(
+            clause.getVariableName(),
+            clause.getSequenceType(),
+            clause.getMetadata(),
+            clause.getVariableHighestStorageMode(this.visitorConfig)
+        );
 
-        StaticContext result = visitFlowrVarDeclaration(expression, argument);
-        if (expression.getPositionalVariableReference() != null) {
+        if (clause.getPositionalVariableName() != null) {
             result.addVariable(
-                expression.getPositionalVariableReference().getVariableName(),
+                clause.getPositionalVariableName(),
                 new SequenceType(ItemType.integerItem),
-                expression.getMetadata(),
+                clause.getMetadata(),
                 ExecutionMode.LOCAL
             );
         }
         return result;
     }
 
     @Override
-    public StaticContext visitLetClauseVar(LetClauseVar expression, StaticContext argument) {
-        this.visit(expression.getExpression(), argument);
-        expression.initHighestExecutionAndVariableHighestStorageModes(this.visitorConfig);
-        return visitFlowrVarDeclaration(expression, argument);
+    public StaticContext visitLetClause(LetClause clause, StaticContext argument) {
+        // TODO visit at...
+        this.visit(clause.getExpression(), argument);
+        clause.initHighestExecutionMode(this.visitorConfig);
+
+        StaticContext result = new StaticContext(argument);
+        // TODO for now we only support as/default, no inference, flags
```

Comment:
Can we extract these TODOs to issues?

Reply:
I think this was an overlook, there is nothing else than as for let clauses.